opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -325,6 +331,161 @@ namespace Opm
     }
 
 
+    template<typename TypeTag>
+    template<class ValueType>
+    ValueType
+    WellInterface<TypeTag>::
+    calculateBhpFromThp(const std::vector<ValueType>& rates,
```

Comment:
For MSW, the top segment might not have a perforation. Unless we have some strong evidence, we should not assume to do THP related to be the same way with StandardWell.  It should be similar though. 

As we discussed, we should only address the calculation related to `BHP` limit in this PR. If we handle THP related, it will be a much more extensive PR. 

It will be good to just through a warning when handle well potential related to THP limit, including in StandardWell. 

Referring to the discussion in https://github.com/OPM/opm-simulators/issues/1607 , we should address THP related in a different PR.

Reply:
I have added a warning if THP limit if the user tries to compute well potentials for cases with THP and MSW. (line 593 in MultisegmentWell_impl.hpp) 

Even though some details may be different, I would assume that most of the code will be common for MSW and standard wells regarding VFP. But since you are planning to do the VPF stuff I suggest that things gets merged as is, and you can refactor and change the code according to your need later. It is not easy to guess for me to guess what may or may not be a good code structure.